### PR TITLE
[DevTools] Don't assert on child order changes below a filtered bailout

### PR DESCRIPTION
**Pull request by hoxyq**
**Setup**

- Component filter: `<Suspense>` hidden, host elements visible.
- `Reader` calls `use()` on a promise that never resolves, so any boundary
  containing it shows its fallback.

```jsx
<div>
  {extra ? <em key="s">side</em> : null}
  <A key="a" collapsed={collapsed} />
</div>
```

```jsx
// A collapsed={false}                    // A collapsed={true}
<Suspense fallback="a-outer-fb">          <Suspense fallback="a-fb">
  <Suspense fallback="a-inner-fb">          <Reader />
    <Reader />                            </Suspense>
  </Suspense>
  <Suspense fallback="a-sib-fb">
    <div>a-sibling</div>
  </Suspense>
</Suspense>
```

---

## Commit 1 — `collapsed={false} extra={false}`

React. A `<Suspense>` renders its content into an `Offscreen`; when it suspends,
the fallback is added as a **sibling** of that `Offscreen` and the content is
hidden rather than unmounted.

```
<div>
└── <A key="a">
    └── Suspense·outer                          not suspended
        └── Offscreen                           visible
            ├── Suspense·inner                  SUSPENDED
            │   ├── Offscreen                   hidden
            │   │   └── <Reader>                still mounted
            │   └── <div>a-inner-fb</div>       fallback
            └── Suspense·sib                    not suspended
                └── Offscreen                   visible
                    └── <div>a-sibling</div>
```

Store. `Suspense` and `Offscreen` are filtered, so surviving descendants hoist to
the nearest visible ancestor, `<A>`.

```
[root]
└── <App>
    └── <div>
        └── <A key="a">
            ├── <div>          a-inner-fb
            └── <div>          a-sibling
```

Identical before and after the fix.

---

## Commit 2 — `collapsed={true} extra={true}`

Two changes land in one commit: `<em>` mounts, and `A` restructures so that
`Suspense·outer` now directly contains `<Reader>` and therefore suspends.

React, after the commit:

```
<div>
├── <em key="s">                               NEW, id 7
└── <A key="a">
    └── Suspense·outer                         now SUSPENDED
        ├── Offscreen                          hidden
        │   └── <Reader>
        └── <div>a-fb</div>                    fallback
```

While reconciling, DevTools reaches the `Offscreen` of the outgoing
`Suspense·sib`. React handed that node back with its child pointer untouched, so
DevTools takes its "children unchanged" shortcut — but the walk beneath reports
that the panel's rows changed, because the subtree around it was restructured:

```
Offscreen (of Suspense·sib)     ← filtered, so it has no Store row of its own
    sameChildPtr = true         ← locally nothing changed
    flagSet      = true         ← the panel's rows did change
        └── throw "The children should not have changed if we pass in the same set."
```

React catches whatever DevTools throws and logs it at most once, so the app is
unaffected and nothing surfaces. But `flushPendingEvents` never runs, and the
`add <em id=7>` recorded earlier in this same commit is left unsent in
`pendingOperations`.

Store — **before the fix** (unchanged from commit 1, now stale):

```
[root]
└── <App>
    └── <div>
        └── <A key="a">
            ├── <div>          stale, no longer exists
            └── <div>          stale, no longer exists
                               <em> never arrived
```

Store — **after the fix**:

```
[root]
└── <App>
    └── <div>
        ├── <em key="s">
        └── <A key="a">
            └── <div>          a-fb
```

---

## Commit 3 — `collapsed={true} extra={false}`

`<em>` unmounts. This commit walks cleanly and does flush — but the batch still
carries commit 2's leftovers, and `flushPendingEvents` writes **all removals to
the front**, ahead of everything else:

```
[ remove <div> ][ remove <div> ][ remove <em id=7> ]   ← hoisted to the front
[ add <em id=7> ][ add <div>a-fb ]                     ← never reached
```

The Store applies the two removals, then hits a removal for id 7 — a node it was
never told about — throws, and abandons the rest of the array.

Store — **before the fix**:

```
[root]
└── <App>
    └── <div>
        └── <A key="a">        children removed, nothing added back
```

```
Cannot remove node "7" because no matching node was found in the Store.
```

From here the Store never applies another operation.

Store — **after the fix**:

```
[root]
└── <App>
    └── <div>
        └── <A key="a">
            └── <div>          a-fb
```

---

## Why the check was wrong

Suspending **hides** content instead of unmounting it, so state survives a
reveal. That means React can reuse every fiber while the set of rows the panel
should display changes completely. `nextFiber.child === prevFiber.child` is a
fact about one link; it says nothing about the subtree below it.

The deleted assertion read that local fact as proof of a global one:

| what the code knew            | what it concluded              |
| ----------------------------- | ------------------------------ |
| React reused the same objects | the panel shows the same rows  |

Those come apart precisely when a `Suspense` or `Activity` boundary changes
visibility — which is the one case this branch had to handle.

The fix lets the "rows changed" signal propagate to the closest unfiltered
ancestor, which re-sends its child list. That is the path the surrounding code
already uses; twenty lines below the assertion there is a branch whose only
purpose is to forward this flag from a filtered fiber.

### CODE REVIEW

**Review by dianatofficial — COMMENTED** [review]
Great catch on this edge case. Fallback logic is robust.

File: packages/react-devtools-shared/src/__tests__/storeFilteredSuspense-test.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    actAsync = utils.actAsync;
+  });
+
+  afterEach(() => {
+    store.componentFilters = previousComponentFilters;
+  });
+
+  const {render} = getVersionedRenderImplementation();
+
+  function buildApp() {
```
Comment (eps1lon) [inline packages/react-devtools-shared/src/__tests__/storeFilteredSuspense-test.js]: let's inline this to match how we currently write tests.

File: packages/react-devtools-shared/src/__tests__/storeFilteredSuspense-test.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      // Assigning replaces the whole list, so host components stay visible.
+      // The control test below shows both halves are required.
+      store.componentFilters = [
+        utils.createElementTypeFilter(12 /* ElementTypeSuspense */),
+      ];
+    });
+
+    await actAsync(() => render(<App collapsed={false} extra={false} />));
+    await actAsync(() => render(<App collapsed={true} extra={true} />));
+    await actAsync(() => render(<App collapsed={true} extra={false} />));
```
Comment (eps1lon) [inline packages/react-devtools-shared/src/__tests__/storeFilteredSuspense-test.js]: Let's include store snapshots here as well to ensure this is actually showing the right thing.
